internal/beater/beater.go

```diff
@@ -177,6 +177,36 @@ func (s *Runner) Run(ctx context.Context) error {
 		}
 	}
 
+	// Obtain the memory limit for the APM Server process. Certain config
+	// values will be sized according to the maximum memory set for the server.
+	var memLimit float64
+	if cgroupReader := newCgroupReader(); cgroupReader != nil {
+		if limit, err := cgroupMemoryLimit(cgroupReader); err != nil {
+			s.logger.Warn(err)
+		} else {
+			// Limit the memory to 80% of the cgroup memory limit.
+			memLimit = float64(limit) / 1024 / 1024 / 1024 * 0.8
+		}
+	}
+	if memLimit <= 0 {
+		s.logger.Info("no cgroups detected, falling back to total system memory")
+		if limit, err := systemMemoryLimit(); err != nil {
+			s.logger.Warn(err)
+		} else {
+			// If no cgroup limit is set, only return 50% of the total memory.
+			// to have a margin of safety for other processes.
+			memLimit = float64(limit) / 1024 / 1024 / 1024 * 0.5
+		}
+	}
+	if memLimit > 0 {
```

Comment:
Also, perhaps only call maxConcurrentDecoders (probably renamed to include "memory" in the name somewhere) in case MaxConcurrentDecoders hasn't been explicitly specified. The log message below is a bit misleading otherwise, as it's not based on the memory.